### PR TITLE
Solved a task, implemented all additional functions. Didn't implement whole functions of dict

### PR DESCRIPTION
**Pull request by ORENYT**
I used special functions to implement default dict functional, also used additional tests for delete object.

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -1,2 +1,130 @@
...
+        except KeyError:
+            return -1
+
+    def update(self, second_dict: "Dictionary") -> None:
+        nodes = [
+            node
+            for node in second_dict.__dict__["_Dictionary__hashtable"]
+            if node is not None
+        ]
+        for node in nodes:
```
Comment (denys-source) [inline app/main.py:68]: I would recommend you using one loop like that:
```
for node in second_dict.__dict__["_Dictionary__hashtable"]:
    if not is not None:
        self.__setitem__(node.key, node.value)
```

File: app/main.py [hunk 2]
```diff
@@ -1,2 +1,130 @@
...
+        preferred_cell = node.key_hash % length
+        for i in range(preferred_cell, length):
+            if self.__hashtable[i] and self.__hashtable[i].key == node.key:
+                self.__hashtable[i] = node
+                return
+            if self.__hashtable[i] is None:
+                self.__hashtable[i] = node
+                return
+        for i in range(0, preferred_cell):
+            if self.__hashtable[i] and self.__hashtable[i].key == node.key:
```
Comment (denys-source) [inline app/main.py:125]: I think one `if` condition with `or` operator will remove duplicate code

**Review by MarianKovalyshyn — APPROVED** [review]
I think you could replace these loops with one loop because they are very similar and they can confuse at first look. But everything else looks good to me.
```
for i in range(preferred_cell, length):
    if self.__hashtable[i] and self.__hashtable[i].key == key:
        return self.__hashtable[i].value
for i in range(0, preferred_cell):
    if self.__hashtable[i] and self.__hashtable[i].key == key:
        return self.__hashtable[i].value
 ```

File: app/main.py [hunk 3]
```diff
@@ -1,2 +1,130 @@
...
+        self.__delitem__(key)
+        return item
+
+    def __check_overload(self) -> None:
+        """
+        Do not use this function, it calculates when hashtable should
+        be multiplied automatically
+        """
+        length = len(self.__hashtable)
+        if length - self.__hashtable.count(None) > length * self.__load_factor:
```
[the remark this answers is not in the document]
Comment (ORENYT) [reply app/main.py:98]: PM'ed you, what do you actually think it is better to change? I use length to calculate full length of dict, but __len__ calculates non-None elements

File: app/main.py [hunk 4]
```diff
@@ -1,2 +1,123 @@
+import dataclasses
+from typing import Hashable
+
+
+@dataclasses.dataclass
+class Node:
+    key: object
```
Comment (Kroll410) [inline app/main.py:7]: Hashable type

File: app/main.py [hunk 5]
```diff
@@ -1,2 +1,123 @@
...
+    key_hash: int
+    value: object
```
Comment (Kroll410) [inline app/main.py:9]: Any type

File: app/main.py [hunk 6]
```diff
@@ -1,2 +1,123 @@
...
+
+
 class Dictionary:
-    pass
+    __load_factor = 2 / 3
+
+    def __init__(self) -> None:
+        self.__hashtable = [None] * 8
+
+    def __setitem__(self, key: Hashable, value: object) -> None:
```
Comment (Kroll410) [inline app/main.py:18]: `object` is too broad and high-abstract type. Use typings

File: app/main.py [hunk 7]
```diff
@@ -1,2 +1,123 @@
...
+                self.index += 1
+            return (
+                self.__hashtable[self.index].key,
+                self.__hashtable[self.index].value,
+            )
+        except IndexError:
+            raise StopIteration
+
+    def pop(self, key: Hashable) -> object:
+        item = self.__getitem__(key)
```
Comment (Kroll410) [inline app/main.py:85]: if not found, default value should be returned
image

File: app/main.py [hunk 8]
```diff
@@ -1,2 +1,123 @@
...
+
+    def update(self, second_dict: "Dictionary") -> None:
+        for node in second_dict.__dict__["_Dictionary__hashtable"]:
+            if node:
+                self.__setitem__(node.key, node.value)
+            self.__setitem__(node.key, node.value)
+
+    def __iter__(self) -> list:
+        self.index = 0
+        return self.__hashtable
```
Comment (Kroll410) [inline app/main.py:70]: `dict` when being iterated sends us the keys, if not values or items methods were specified

**Review by Kroll410 — CHANGES_REQUESTED** [review]

File: app/main.py (hunk 8 above)
Comment (ORENYT) [reply app/main.py:70]: Understood, accepted.

File: app/main.py [hunk 9]
```diff
@@ -1,2 +1,125 @@
...
+        for i in range(preferred_cell, length):
+            if self.__hashtable[i] and self.__hashtable[i].key == key:
+                return self.__hashtable[i].value
+        for i in range(0, preferred_cell):
+            if self.__hashtable[i] and self.__hashtable[i].key == key:
+                return self.__hashtable[i].value
+        raise KeyError("Key not found")
+
+    def __len__(self) -> int:
+        return len([node for node in self.__hashtable if node is not None])
```
[the remark this answers is not in the document]
Comment (ORENYT) [reply app/main.py:37]: Oh, sorry, fixed

File: app/main.py [hunk 10]
```diff
@@ -1,2 +1,125 @@
...
+        preferred_cell = hashed_key % length
+        for i in range(preferred_cell, length):
+            if self.__hashtable[i] and self.__hashtable[i].key == key:
+                self.__hashtable[i] = None
+                return
+        for i in range(0, preferred_cell):
+            if self.__hashtable[i] and self.__hashtable[i].key == key:
+                self.__hashtable[i] = None
+                return
+        raise KeyError("Key not found")
```
[the remark this answers is not in the document]
Comment (ORENYT) [reply app/main.py:54]: I swear I fixed it yesterday, omg. Something wrong with github i think

File: app/main.py [hunk 11]
```diff
@@ -1,2 +1,127 @@
...
+                self.__hashtable[i] = None
+                self.length -= 1
+                return
+        raise KeyError("Key not found")
+
+    def get(self, key: object, default: Any = -1) -> object:
+        try:
+            return self.__getitem__(key)
+        except KeyError:
+            return default
```
[the remark this answers is not in the document]
Comment (ORENYT) [reply app/main.py:60]: Changed to None

File: app/main.py [hunk 12]
```diff
@@ -1,2 +1,127 @@
...
+
+    def update(self, second_dict: "Dictionary") -> None:
```
[the remark this answers is not in the document]
Comment (ORENYT) [reply app/main.py:62]: Deleted this method

File: app/main.py [hunk 13]
```diff
@@ -1,2 +1,127 @@
...
+        for node in second_dict.__dict__["_Dictionary__hashtable"]:
+            if node:
+                self.__setitem__(node.key, node.value)
+            self.__setitem__(node.key, node.value)
```
[the remark this answers is not in the document]
Comment (ORENYT) [reply app/main.py]: this method is removed as an optional

**Review by ORENYT — COMMENTED** [review]
Accepted all changes

**Pull request closed without merging**